type stable last for heterogeneous tuples

Sacha0 opened the pull request:
`last` is not type stable for heterogeneous tuples. This pull request provides a specialization of `last` that is type stable for heterogeneous tuples. (Should something like [the trick used to limit code growth in the definition of `map` for tuples](https://github.com/JuliaLang/julia/blob/374c3d6b3c3e15c00f7d3df8b5cb7c8a763aa746/base/tuple.jl#L110) limit inlining here as well?). Best!

JeffBezanson commented:
`t[nfields(t)]` might work here, since we can infer a constant for `nfields`.

Sacha0 commented:
> t[nfields(t)] might work here, since we can infer a constant for nfields.

Unfortunately
```julia
julia> foo(t) = t[nfields(t)]
foo (generic function with 1 method)

julia> @code_warntype foo((1, 2.))
Variables:
  #self#::#foo
  t::Tuple{Int64,Float64}

Body:
  begin
      SSAValue(0) = (Main.nfields)(t::Tuple{Int64,Float64})::Int64
      return (Base.getfield)(t::Tuple{Int64,Float64},SSAValue(0))::Union{Float64,Int64}
  end::Union{Float64,Int64}
```
though
```julia
julia> foo{N}(t::NTuple{N}) = t[N]
foo (generic function with 1 method)

julia> @code_warntype foo((1, 2.0))
Variables:
  #self#::#foo
  t::Tuple{Int64,Float64}

Body:
  begin
      return (Base.getfield)(t::Tuple{Int64,Float64},$(Expr(:static_parameter, 1)))::Float64
  end::Float64
```
Thoughts? Thanks!

JeffBezanson commented:
Ah, the extra call to getindex could be the problem. Maybe `getfield(t, nfields(t))`. We should add special cases for tuple getindex and length.

Sacha0 commented:
> Maybe getfield(t, nfields(t)).

You mean like so?
```julia
julia> foo(t::Tuple) = getfield(t, nfields(t))
foo (generic function with 1 method)

julia> @code_warntype foo((1, 2.0))
Variables:
  #self#::#foo
  t::Tuple{Int64,Float64}

Body:
  begin
      return (Main.getfield)(t::Tuple{Int64,Float64},(Main.nfields)(t::Tuple{Int64,Float64})::Int64)::Union{Float64,Int64}
  end::Union{Float64,Int64}
```
Best!

martinholters commented:
Like so:
```Julia
julia> foo{T<:Tuple}(t::T) = getfield(t, nfields(T))
foo (generic function with 1 method)

julia> @code_warntype foo((1, 2.0))
Variables:
  #self#::#foo
  t::Tuple{Int64,Float64}

Body:
  begin 
      return (Main.getfield)(t::Tuple{Int64,Float64},(Main.nfields)($(Expr(:static_parameter, 1)))::Int64)::Float64
  end::Float64
```

mbauman commented:
Or even just:

```jl
julia> Base.length{T<:Tuple}(t::T) = nfields(T)

julia> @code_warntype last((1,2.))
Variables:
  #self#::Base.#last
  a::Tuple{Int64,Float64}

Body:
  begin
      SSAValue(0) = (Main.nfields)(Tuple{Int64,Float64})::Int64
      return (Base.getfield)(a::Tuple{Int64,Float64},SSAValue(0))::Float64
  end::Float64
```

Sacha0 committed:
Improve length (and thereby endof and last) for tuples.

Make length(/endof) for tuples easier for the compiler to reason about. last for heterogeneous tuples becomes type stable as a result.

Sacha0 commented:
Beautiful! Pull request updated with `@mbauman`'s suggestion. Thanks all!

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19540#issuecomment-266884637) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ab99df501ef390247b5da7d11d3619c4dd4852df/76586de_vs_66ba9ca/report.md). cc @jrevels

Sacha0 commented:
Nanosoldier thoroughly disapproves of that variation on `length` unfortunately. Thoughts? Best!

martinholters commented:
Looks like the `length` defined in this PR is inferior if the exact type of the `Tuple` cannot be inferred. E.g. with
```Julia
foo(n) = length(((1:n)...)) # argument to length is inferred as Tuple{Vararg{Int64,N}}
```
I have on master:
```Julia
BenchmarkTools.Trial: 
  memory estimate:  256.00 bytes
  allocs estimate:  4
  --------------
  minimum time:     380.678 ns (0.00% GC)
  median time:      384.134 ns (0.00% GC)
  mean time:        404.979 ns (3.64% GC)
  maximum time:     7.432 μs (91.05% GC)
  --------------
  samples:          10000
  evals/sample:     202
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
With this PR:
```Julia
BenchmarkTools.Trial: 
  memory estimate:  320.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     651.307 ns (0.00% GC)
  median time:      665.187 ns (0.00% GC)
  mean time:        689.331 ns (2.87% GC)
  maximum time:     9.572 μs (90.97% GC)
  --------------
  samples:          8628
  evals/sample:     166
  time tolerance:   15.00%
  memory tolerance: 1.00%
```

I have no idea how to proceed, though. Only changing `last` might be a safe choice, but could also leave some potential unused for other users of `length`...

JeffBezanson commented:
The version with `getfield(t, nfields(t))` ought to work; looking at it again I discovered the `nfields` t-function was not as aggressive as I thought. Try this patch:

```diff
diff --git a/base/inference.jl b/base/inference.jl
index b57f521..cdf57d1 100644
--- a/base/inference.jl
+++ b/base/inference.jl
@@ -382,6 +382,8 @@ add_tfunc(isdefined, 1, IInf, (args...)->Bool)
 add_tfunc(Core.sizeof, 1, 1, x->Int)
 add_tfunc(nfields, 1, 1, (x::ANY)->(isa(x,Const) ? Const(nfields(x.val)) :
                              isType(x) && isleaftype(x.parameters[1]) ? Const(nfields(x.par
+                             !isType(x) && isleaftype(x) ? Const(nfields(x)) :
+                             isa(x,DataType) && x.name===Tuple.name && !isvatuple(x) ? Cons
                              Int))
 add_tfunc(Core._expr, 1, IInf, (args...)->Expr)
 add_tfunc(applicable, 1, IInf, (f::ANY, args...)->Bool)
```

Sacha0 commented:
Superseded by #19605. Thanks `@JeffBezanson`!
